quelpa.el

```diff
@@ -1806,14 +1864,25 @@ nil."
   (when (quelpa-setup-p) ;if init fails we do nothing
     (let* ((quelpa-upgrade-p (if current-prefix-arg t quelpa-upgrade-p)) ;shadow `quelpa-upgrade-p'
            (quelpa-stable-p quelpa-stable-p) ;shadow `quelpa-stable-p'
-           (cache-item (if (symbolp arg) (list arg) arg)))
+           (quelpa-force-upgrade-p quelpa-force-upgrade-p) ;shadow `quelpa-force-upgrade-p'
```

Comment:
Why is this variable being shadowed?  That's sort of a "code smell", if you know what I mean.  :)

Reply:
To prevent someone down the line to modify the global `quela-force-upgrade-p`?
I saw the two previous line is also adopting this shadowing practice, so I just follow it.

Reply:
I guess you could use a separate var and then check for that in the places it is used. Shadowing makes this unnecessary since one just overwrites the default but the default still applies if it is not shadowed.

Reply:
IIUC there seems to be redundancy and/or conflict between `quelpa-force-upgrade-p` and `quelpa--use-current-state-p`.  Aren't these two variables supposed to govern the same thing, namely whether uncommitted changes in the local clone are discarded before building?

Reply:
It's a little bit different, the `quelpa-force-upgrade-p` is using to forcefully upgrade the package. I.e: using `git reset --hard`, it will even delete the directory to make the upgrade sucessfull.
Without `quelpa-force-upgrade-p` and `quelpa--use-current-state-p`, now Quelpa will simply using checkout, which can fail if there's non commit change in the package.
And finally, `quelpa--use-current-state-p` is internal variable, it indicate that we don't want to checkout from `origin/master` but simple keep current working state and try doing upgrade from there.

So they govern 2 different aspects, whether not pushed changes get keep and whether uncommitted changes can stop upgrading.

Here is the table:

|                                | quelpa-force-upgrade-p: t             | quelpa-force-upgrade-p: f                                   |
|--------------------------------|---------------------------------------|-------------------------------------------------------------|
| quelpa--use-current-state-p: t | Discard uncommitted, unpushed changes | Discard nothing, use current working                        |
| quelpa--use-current-state-p: f | Discard uncommitted, unpushed changes | Checkout remote master, fail if there's uncommitted changes |

Reply:
I think this is too complicated.  If we have to use a matrix to understand how these options are intended to interact when we're writing and reviewing the code, how can we expect users to understand them when using it?

I think it should work more like this when building a package:

a.  If a variable `quelpa-build-target` is `master`, attempt to checkout `master`.  If it fails (e.g. due to uncommitted changes), signal a helpful error and let the user resolve it.  If it works, build the package.
b.  If `quelpa-build-target` is `HEAD`, build the package with the current HEAD.

If absolutely necessary (and I'm skeptical that it is, but we should talk about it), a `(defvar quelpa-force-checkout)` or `(defvar quelpa-reset-hard)` variable (not a customization option) could be used with an interactive prefix to force option A above to reset to `master` if the checkout fails--in which case, a warning should probably be displayed to ensure the user realizes that the hard reset was required.

By doing it this way, we avoid the confusing double-negatives caused by multiple booleans interacting, and with the variable we expose to the user, we use Git terms rather than true/false, which makes its behavior easier to understand: build from `master` or build from `HEAD`.

What do you think?

Reply:
It's great, perhaps we can do this in separate PR?
I like the part about showing warnings, we definitely should do that.

For now if we want `quelpa-build-target`, we may also need to define it's using which remote (origin, upstream or whatever), which is kind of hidden from user. That why I simplify that into just using working state (HEAD) or not (origin/master).

About the force upgrade option, I think that's necessary. It gives user a smooth upgrading without any interacting why waiting for package to upgrade (surfing web maybe). I can make that one back to variable instead of customization option though. So we can ensure user only upgrading smoothly when using `quelpa-upgrade-all`.
Actually that's one of the reason I switch back to `quelpa` from `straight.el`. In `quelpa` when I don't care about package developing, I can just always get the latest version even if my git repo somehow become dirty (file permission, CRLF auto convert and such).

What do you think?

Reply:
> For now if we want quelpa-build-target, we may also need to define it's using which remote (origin, upstream or whatever), which is kind of hidden from user. That why I simplify that into just using working state (HEAD) or not (origin/master).

If necessary, we could do something like `(defcustom quelpa-build-target-remote "origin" ...)` so users can override the default.  I think that would be simple to add.

> About the force upgrade option, I think that's necessary. It gives user a smooth upgrading without any interacting why waiting for package to upgrade (surfing web maybe). I can make that one back to variable instead of customization option though. So we can ensure user only upgrading smoothly when using quelpa-upgrade-all.

> Actually that's one of the reason I switch back to quelpa from straight.el. In quelpa when I don't care about package developing, I can just always get the latest version even if my git repo somehow become dirty (file permission, CRLF auto convert and such).

In that case, I'd suggest that forcing a hard reset should be done with a prefix argument to the interactive command, and that it should also ask for confirmation with `yes-or-no-p`.  I think Quelpa should be very careful not to discard uncommitted changes unless the user *really* wants to.

> It's great, perhaps we can do this in separate PR?  

Well, yes, it would be easier to have these discussions and make these decisions if the PR were more narrowly focused.  For example, looking at the first message, which describes the changes included in this PR, most of those changes are orthogonal to each other.  If they were in separate PRs, each one could be discussed and merged on its own merits without affecting the others.  Then, for example, the upgrade-all command could be added separately from issues about which directory to build in, whether to build from master or HEAD or origin or upstream, etc.

This is the problem with lumping many unrelated changes into a single PR.  You did split some of them into that bsdtar-related one, but this one is still rather monolithic.  :)

Reply:
Recently there was a merge request that added the possibility to choose the remote: https://framagit.org/steckerhalter/quelpa/issues/166
`melpa` recipes let you specify the commit and/or branch and now also the remote since this merge. I'm not sure if this conflicts with some of the suggestions/changes here. It would be good if we could add some tests that ensure we do not break backwards compatibility or if we do that we know and have a clear strategy.

Reply:
Yes, I think this concern needs to be addressed before this PR should be merged.

Reply:
This change has been split into another PR (#156), we can continue to discuss over there.